packages/pkl.lua/tests/parser.pkl-expected.pcf

```diff
@@ -0,0 +1,232 @@
+    new { foo = 1 }
+    new { foo { a = 2; b = 2 } }
+    new { a = 2; foo { a = 1 } }
+    new { a = 2; [true] = 1 }
+    new { a = 1; b = 1; [true] = 2 }
+    new { ["foo"] = new Mapping { ["a"] = 1; ["b"] = 1; [true] = 2 } }
+  }
+  ["path converters apply after converting keys"] {
+    new { x { [42] = true } }
+    new { x { `42` = "matched: true" } }
+    new { x { `42` = "matched: true" } }
+  }
+  ["path converters in tables as table keys"] {
+    // the use of amending here fixes an odd ordering issue
+    // Dynamic literals sort properties before keys, so amending preserves the expected order
+    fixupTableKeys((new Dynamic { [new Dynamic { a = 1 }] = 1 }) { done = true })
+    fixupTableKeys((new Dynamic { [new Dynamic { a = 2; b = 1 }] = 1 }) { a = 2 })
+    fixupTableKeys((new Dynamic { [new Dynamic { a = 1; b = 1 }] = 1 }) { a = 2 })
+    fixupTableKeys(new Dynamic { foo { [new Dynamic { a { b = 2 }; b = 1 }] = 1 } })
+  }
+  ["converters can return non-Lua types"] {
+    Pair("foo", 1)
+    new { foo = Pair("int", 1) }
+    new { [Pair("int", 5)] = true }
+    new { foo { [Pair("int", 5)] = true } }
+    new Mapping { [Pair("int", 5)] = true }
+    new Mapping { ["foo"] = new Mapping { [Pair("int", 5)] = true } }
+  }
+  ["fixtures"] {
+    fixupTableKeys(new Dynamic {
+      greeting = "Hello, world!"
+      snippet = """
+        Long Lua strings can be written inside brackets like this: [[
+        This is a multiline string.
+        It ends at the close double-bracket.]]
+
+        """
+      hex_floats {
+        16.0
+        0.9999990463256836
+        -400.0
+      }
+      tableKeys {
+        identifier = true
+        string = "yes"
+        [42] = "the meaning of life"
+        [3.14159] = "pi"
+        [true] = "very true"
+        [new Dynamic { 1; 2; 3 }] = "even tables can be keys"
+      }
+    })
+    new Dynamic {
+      foo = "bar"
+      count = 2
+      enable = true
+      frob = null
+      ports { 80; 443 }
+      ips {
+        localhost = "127.0.0.1"
+        `example.com` = "93.184.215.14"
+      }
+      ` ` = "space"
+    }
+  }
+}
+
+// See this same function at the end of parser.pkl for an explanation.
+local function fixupTableKeys(value: Dynamic|Mapping|Listing): Dynamic|Mapping|Listing =
+  if (value is Listing)
+    value.toList().map((v) ->
+      if (v is Dynamic|Mapping|Listing) fixupTableKeys(v)
+      else v
+    ).toListing()
+  else
+    let (map = value.toMap().map((k,v) ->
+      let (key =
+        if (k is Dynamic) new PcfRenderDirective { before = "Dynamic "; value = k }
+        else if (k is Mapping) new PcfRenderDirective { before = "Mapping "; value = k }
+        else if (k is Listing) new PcfRenderDirective { before = "Listing "; value = k }
+        else k)
+        if (v is Dynamic|Mapping|Listing) Pair(key, fixupTableKeys(v))
+        else Pair(key,v)
+    ))
+      if (value is Dynamic) (map.toDynamic()) {
+        ...value.toList()
+      } else map.toMapping()
```

Comment:
While I see the use-case, this isn't valid PCF. This happens to work right now, but it will break when we start using an actual PCF parser for running tests.

I think the best way to write these tests, for now, is to use the PCF renderer for generating examples.

If/when we have an in-language Pkl renderer, that will make these tests even better because you will get output like `foo = new Dynamic {}`.

Reply:
This code here is necessary precisely because the Pcf renderer produces output that cannot be parsed. As long as the tests all pass it's fine, but as soon as a test fails then Pkl needs to be able to parse the `parser.pkl-actual.pcf` file. This means I need to muck with the output in the test file, which causes it to render something that is parseable, but then it fails the equality comparison (since it's now comparing a value that includes `PcfRenderDirective` with something that doesn't). So I need the same code in the pcf file to modify the expected value to have the same `PcfRenderDirective` values.

I suppose I could do a different hack that replaces the Dynamic used as a key with e.g. a string containing the rendered Pcf. Or I could just write the `PcfRenderDirective` values inline in the output, but then that raises the possibility of `fixupTableKeys()` producing ordering differences in the output.

Is this a blocker?

Reply:
Done. The string approach worked fine.